src/NuGet.Core/NuGet.ProjectModel/Utf8JsonStreamReader.cs

```diff
@@ -0,0 +1,338 @@
+
+        internal Utf8JsonStreamReader(Stream stream)
+        {
+            if (stream is null)
+            {
+                throw new ArgumentNullException(nameof(stream));
+            }
+
+            _complete = false;
+            _stream = stream;
+            _buffer = ArrayPool<byte>.Shared.Rent(1024);
+            _stream.Read(_buffer, 0, 3);
+            var offset = 0;
+            if (!_utf8Bom.SequenceEqual(_buffer.AsSpan(0, 3)))
+            {
+                offset = 3;
+            }
+            _stream.Read(_buffer, offset, _buffer.Length - offset);
+            _reader = new Utf8JsonReader(_buffer, isFinalBlock: false, state: new JsonReaderState(new JsonReaderOptions
+            {
+                AllowTrailingCommas = true,
+                CommentHandling = JsonCommentHandling.Skip,
+            }));
+            _reader.Read();
+        }
+
+        internal bool IsFinalBlock => _reader.IsFinalBlock;
+
+        internal JsonTokenType TokenType => _reader.TokenType;
+
+        internal int BufferSize()
+        {
+            if (_complete)
+            {
+                throw new InvalidOperationException("Cannot read from completed Utf8JsonStreamReader");
+            }
+            return _buffer.Length;
+        }
+
+        internal bool ValueTextEquals(ReadOnlySpan<byte> utf8Text) => _reader.ValueTextEquals(utf8Text);
+
+        internal bool ValueTextEquals(string text) => _reader.ValueTextEquals(text);
+
+        internal bool TryGetInt32(out int value) => _reader.TryGetInt32(out value);
+
+        internal string GetString() => _reader.GetString();
+
+        internal bool GetBoolean() => _reader.GetBoolean();
+
+        internal int GetInt32() => _reader.GetInt32();
+
+        internal void CompleteRead()
+        {
+            _complete = true;
+            ArrayPool<byte>.Shared.Return(_buffer);
+        }
+
+        internal bool Read()
+        {
+            if (_complete)
+            {
+                throw new InvalidOperationException("Cannot read from completed Utf8JsonStreamReader");
+            }
+
+            bool wasRead;
+            while (!(wasRead = _reader.Read()) && !_reader.IsFinalBlock)
+            {
+                GetMoreBytesFromStream();
+            }
+            return wasRead;
+        }
+
+        internal bool TrySkip()
+        {
+            if (_complete)
+            {
+                throw new InvalidOperationException("Cannot read from completed Utf8JsonStreamReader");
+            }
+
+            bool wasSkipped;
+            while (!(wasSkipped = _reader.TrySkip()) && !_reader.IsFinalBlock)
+            {
+                GetMoreBytesFromStream();
+            }
+            return wasSkipped;
+        }
+
+        //internal IList<T> ReadObjectAsList<T>(Utf8JsonStreamReaderConverter<T> streamReaderConverter)
+        //{
+        //    if (TokenType != JsonTokenType.StartObject)
+        //    {
+        //        return Array.Empty<T>();
```

Comment:
https://github.com/NuGet/NuGet.Client/pull/5529/files#r1414981447
> Array.Empty(); - @davkean 

Done for both instances